packages/docregistry/src/default.ts

```diff
@@ -383,3 +399,213 @@ abstract class ABCWidgetFactory<T extends DocumentRegistry.IReadyWidget, U exten
   private _defaultFor: string[];
   private _widgetCreated = new Signal<DocumentRegistry.IWidgetFactory<T, U>, T>(this);
 }
+
+
+
+/**
+ * A widget for rendered mimetype.
+ */
+export
+class MimeRenderer extends Widget implements IRenderMime.IReadyWidget {
+  /**
+   * Construct a new markdown widget.
+   */
+  constructor(options: MimeRenderer.IOptions) {
+    super();
+    this.addClass('jp-MimeRenderer');
+    let layout = this.layout = new PanelLayout();
+    let toolbar = new Widget();
+    toolbar.addClass('jp-Toolbar');
+    layout.addWidget(toolbar);
+    let context = options.context;
+    this.title.label = context.path.split('/').pop();
+    this._rendermime = options.rendermime;
+    this._rendermime.resolver = context;
+    this._context = context;
+    this._mimeType = options.mimeType;
+
+    context.pathChanged.connect(this._onPathChanged, this);
+
+    this._context.ready.then(() => {
+      if (this.isDisposed) {
+        return;
+      }
+      this._render();
+      this._ready.resolve(undefined);
+
+      // Throttle the rendering rate of the widget.
+      this._monitor = new ActivityMonitor({
+        signal: context.model.contentChanged,
+        timeout: options.renderTimeout
+      });
+      this._monitor.activityStopped.connect(this.update, this);
+    });
+  }
+
+  /**
+   * The markdown widget's context.
+   */
+  get context(): DocumentRegistry.Context {
+    return this._context;
+  }
+
+  /**
+   * A promise that resolves when the widget is ready.
+   */
+  get ready(): Promise<void> {
+    return this._ready.promise;
+  }
+
+  /**
+   * Dispose of the resources held by the widget.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    if (this._monitor) {
+      this._monitor.dispose();
+    }
+    super.dispose();
+  }
+
+  /**
+   * Handle `'activate-request'` messages.
+   */
+  protected onActivateRequest(msg: Message): void {
+    this.node.tabIndex = -1;
+    this.node.focus();
+  }
+
+  /**
+   * Handle an `update-request` message to the widget.
+   */
+  protected onUpdateRequest(msg: Message): void {
+    this._render();
+  }
+
+  /**
+   * Render the mime content.
+   */
+  private _render(): void {
+    let context = this._context;
+    let model = context.model;
+    let layout = this.layout as PanelLayout;
+    let data: JSONObject = {};
+    data[this._mimeType] = model.toString();
```

Comment:
This may be the culprit in the Vega extension not working. I think it expects JSON data, rather than a string. We ran into this before, but I don't remember how we resolved it. We probably need to a figure out a way have a clear way of handling MIME types that have JSON or strings for data.